File: Frontend/src/app/service/payment/payment-service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Payment, PaymentIntent, PaymentResponse } from '../../Models/Payment';

@Injectable({
  providedIn: 'root',
})
export class PaymentService {
  private apiUrl = 'http://localhost:8080/api/pagos';

  constructor(private http: HttpClient) {}

  // Crear Payment Intent en Stripe
  createPaymentIntent(amount: number, reservaId: number): Observable<PaymentIntent> {
    return this.http.post<PaymentIntent>(`${this.apiUrl}/create-payment-intent`, {
      amount,
      reservaId,
    });
  }

  // Procesar pago con tarjeta
  processCardPayment(
    paymentIntentId: string,
    reservaId: number,
    amount: number
  ): Observable<PaymentResponse> {
    return this.http.post<PaymentResponse>(`${this.apiUrl}/process-card`, {
      paymentIntentId,
      reservaId,
      amount,
    });
  }

  // Procesar pago en efectivo
  processCashPayment(reservaId: number, amount: number): Observable<PaymentResponse> {
    return this.http.post<PaymentResponse>(`${this.apiUrl}/process-cash`, {
      reservaId,
      amount,
      metodoPago: 'EFECTIVO',
    });
  }

  // Confirmar pago
  confirmPayment(paymentIntentId: string): Observable<PaymentResponse> {
    return this.http.post<PaymentResponse>(`${this.apiUrl}/confirm`, { paymentIntentId });
  }

  // Obtener todos los pagos
  findAll(): Observable<Payment[]> {
    return this.http.get<Payment[]>(this.apiUrl);
  }

  // Obtener pago por ID
  findById(id: number): Observable<Payment> {
    return this.http.get<Payment>(`${this.apiUrl}/${id}`);
  }

  // Obtener pagos por reserva
  findByReserva(reservaId: number): Observable<Payment[]> {
    return this.http.get<Payment[]>(`${this.apiUrl}/reserva/${reservaId}`);
  }

  // Cancelar pago
  cancelPayment(id: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}/cancelar`, {});
  }
}

File: Backend/src/main/java/com/example/lumaresort/service/PaymentService.java
package com.example.lumaresort.service;

import com.example.lumaresort.repository.PaymentRepository;
import com.stripe.Stripe;
import com.stripe.exception.StripeException;
import com.stripe.model.PaymentIntent;
import com.stripe.param.PaymentIntentCreateParams;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

import com.example.lumaresort.entities.Payment;
import com.example.lumaresort.entities.Reserva;
import com.example.lumaresort.repository.ReservaRepository;

@Service
public class PaymentService {

    @Value("${stripe.api.key}")
    private String stripeApiKey;

    private final PaymentRepository paymentRepository;
    private final ReservaRepository reservaRepository;

    public PaymentService(PaymentRepository paymentRepository,
            ReservaRepository reservaRepository) {
        this.paymentRepository = paymentRepository;
        this.reservaRepository = reservaRepository;
    }

    @Transactional
    public PaymentIntent createPaymentIntent(BigDecimal amount, Long reservaId)
            throws StripeException {
        Stripe.apiKey = stripeApiKey;

        // Convertir a centavos
        Long amountInCents = amount.multiply(new BigDecimal(100)).longValue();

        PaymentIntentCreateParams params = PaymentIntentCreateParams.builder()
                .setAmount(amountInCents)
                .setCurrency("cop")
                .addPaymentMethodType("card")
                .putMetadata("reservaId", reservaId.toString())
                .build();

        return PaymentIntent.create(params);
    }

    @Transactional
    public Payment processCardPayment(String paymentIntentId, Long reservaId,
            BigDecimal amount) {
        Reserva reserva = reservaRepository.findById(reservaId)
                .orElseThrow(() -> new RuntimeException("Reserva no encontrada"));

        Payment payment = new Payment();
        payment.setMonto(amount);
        payment.setMetodoPago(Payment.MetodoPago.TARJETA);
        payment.setEstado(Payment.EstadoPago.COMPLETADO);
        payment.setReserva(reserva);
        payment.setStripePaymentIntentId(paymentIntentId);

        return paymentRepository.save(payment);
    }

    @Transactional
    public Payment processCashPayment(Long reservaId, BigDecimal amount) {
        Reserva reserva = reservaRepository.findById(reservaId)
                .orElseThrow(() -> new RuntimeException("Reserva no encontrada"));

        Payment payment = new Payment();
        payment.setMonto(amount);
        payment.setMetodoPago(Payment.MetodoPago.EFECTIVO);
        payment.setEstado(Payment.EstadoPago.PENDIENTE);
        payment.setReserva(reserva);

        return paymentRepository.save(payment);
    }

    public List<Payment> findAll() {
        return paymentRepository.findAll();
    }

    public Payment findById(Long id) {
        return paymentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Pago no encontrado"));
    }

    public List<Payment> findByReserva(Long reservaId) {
        return paymentRepository.findByReservaIdReserva(reservaId);
    }

    @Transactional
    public void cancelPayment(Long id) {
        Payment payment = findById(id);
        payment.setEstado(Payment.EstadoPago.CANCELADO);
        paymentRepository.save(payment);
    }
}
